src/meta/processors/admin/BalancePlan.cpp

```diff
@@ -0,0 +1,198 @@
+        partTasks[std::make_pair(task.spaceId_, task.partId_)].emplace_back(index++);
+    }
+    buckets_.resize(std::min(tasks_.size(), (size_t)FLAGS_task_concurrency));
+    for (auto it = partTasks.begin(); it != partTasks.end(); it++) {
+        size_t minNum = tasks_.size();
+        int32_t i = 0, minIndex = 0;
+        for (auto& bucket : buckets_) {
+            if (bucket.size() < minNum) {
+                minNum = bucket.size();
+                minIndex = i;
+            }
+            i++;
+        }
+        for (auto taskIndex : it->second) {
+            buckets_[minIndex].emplace_back(taskIndex);
+        }
+    }
+}
+
+void BalancePlan::invoke() {
+    status_ = Status::IN_PROGRESS;
+    dispatchTasks();
+    for (size_t i = 0; i < buckets_.size(); i++) {
+        for (size_t j = 0; j < buckets_[i].size(); j++) {
+            auto taskIndex = buckets_[i][j];
+            tasks_[taskIndex].onFinished_ = [this, i, j]() {
+                bool finished = false;
+                {
+                    std::lock_guard<std::mutex> lg(lock_);
+                    finishedTaskNum_++;
+                    if (finishedTaskNum_ == tasks_.size()) {
+                        finished = true;
+                        if (status_ == Status::IN_PROGRESS) {
+                            status_ = Status::SUCCEEDED;
+                        }
+                    }
+                }
+                if (finished) {
+                    CHECK_EQ(j, this->buckets_[i].size() - 1);
+                    saveInStore(true);
+                    onFinished_();
+                } else {
+                    if (j + 1 < this->buckets_[i].size()) {
+                        auto& task = this->tasks_[this->buckets_[i][j + 1]];
+                        task.invoke();
+                    }
+                }
+            };  // onFinished
+            tasks_[taskIndex].onError_ = [this, i, j]() {
+                bool finished = false;
+                {
+                    std::lock_guard<std::mutex> lg(lock_);
+                    finishedTaskNum_++;
+                    status_ = Status::FAILED;
+                    if (finishedTaskNum_ == tasks_.size()) {
+                        finished = true;
+                    }
+                }
+                if (finished) {
+                    CHECK_EQ(j, this->buckets_[i].size() - 1);
+                    saveInStore(true);
+                    onFinished_();
+                } else {
+                    if (j + 1 < this->buckets_[i].size()) {
+                        auto& task = this->tasks_[this->buckets_[i][j + 1]];
+                        task.invoke();
+                    }
+                }
+            };  // onError
+        }  // for (auto j = 0; j < buckets_[i].size(); j++)
+    }  // for (auto i = 0; i < buckets_.size(); i++)
+
+    for (auto& bucket : buckets_) {
+        if (!bucket.empty()) {
+            tasks_[bucket[0]].invoke();
+        }
+    }
+    saveInStore(true);
```

Comment:
saveInStore(true) will be called in onFilish, this is the secend called, why?

Reply:
Because we need to update the balancePlan state